### ISSUE TITLE
Unable to add a custom environment

### ISSUE THREAD

**Issue by gotonickpappas** [shudnal/Seasons#17]
Here is what I did. I copied and pasted "Custom Biome Environments.json" and " Custom environments.json" from the "Default settings" folder to the parent one "shudnal.Seasons". 
Then I opened the Custom environments file and I added a new environment:
[
  {
    "m_cloneFrom": "Mistlands_thunder",
    "m_name": "Clear Summer Ocean",
    "m_isWet": false,
    "m_isFreezing": false,
    "m_isFreezingAtNight": false,
    "m_isCold": true,
    "m_isColdAtNight": true,
    "m_alwaysDark": false
  }
]

I chose to copy from something dramatic for the testing purposes.
I deleted everything else so that it would inherit the values from the default settings of the mod.
Basically my understanding is that the way you set it up is you can clone and then alter a vanilla environment using the Custom environment json file. 
As long as the m_name is unique, you created a new environment with whatever music, sounds, effects etc you wish, and attach it to a biome.

Then I opened the Custom Biome Environments file and I tried to add this environment to the Ocean
{
  "Spring": {
    "add": [],
    "remove": [],
    "replace": []
  },
  "Summer": {
    "add": [
      {
        "m_name": "Ocean",
        "m_environment": {
          "m_environment": "Clear Summer Ocean",
          "m_weight": 1.0,
          "m_ashlandsOverride": false,
          "m_deepnorthOverride": false
        }
      }
    ]
  }
}

Again I deleted everything else.
Now in theory I should be able to open the console type devcommands and then "env Clear Summer Ocean" and I should get mistlands thunder if it is summer and I am on the Ocean. But I do not. 
I checked the log file and there is no line 
[Info : Seasons] Settings updated: Season_Name"
There should be 
[Info : Seasons] Settings updated: Season_Summer", I guess. 

Any idea what is going on?

**Comment by shudnal** [comment]
Do your JSON pass verifying by https://jsoneditoronline.org/ ? You can copy your files one by one and check if there any errors.

```
[Info   :   Seasons] Custom biome environments file loaded
[Info   :   Seasons] Custom environments file loaded
```
and
```
[Info   :   Seasons] Custom biome environments updated
[Info   :   Seasons] Custom environments updated
```
That lines should appear in console if you have copied file in the right place. Which you did apparently. This lines should appear if you change that files.

`[Info : Seasons] Settings updated: Season` appears only on seasonal settings change (Spring.json etc.)

**Comment by shudnal** [comment]
I've just done everything you did and was able to set debug enironment with that name and it worked
![image](https://github.com/shudnal/Seasons/assets/20457922/c337da48-c3b1-4b3d-910e-2c2892d62dfc)

**Comment by shudnal** [comment]
Your understanding looks pretty right.
You can add new custom environments by that mod or any other one adding environments.
And then set how environments should change on every season.

**Comment by shudnal** [comment]
In case my files from ...\BepInEx\config\shudnal.Seasons directory
[Custom Biome Environments.json](https://github.com/user-attachments/files/16119719/Custom.Biome.Environments.json)
[Custom environments.json](https://github.com/user-attachments/files/16119720/Custom.environments.json)

**Comment by shudnal** [comment]
Ah and command **env Clear Summer Ocean** will tell you **Too many parameters** error but it works just fine.

**Comment by gotonickpappas** [comment]
Thank you! Today it is working, and I am suspecting without being sure that it was some missing comma in one of the json files which was resolved with my latest copy and paste. So trying it exactly as I described originally it works. 
I then decided to check if I can assign music to it. 
I have the Custom Audio Mod, which creates a Custom Audio/Music folder and inside that folder you can change the music for biomes or environments as you wish. 
I created a Folder named  LightRainDay (which is the default for light rain if it is day time), I dropped an mp3 in there, and went to test it. 
It works fine.  The code in the Custom Environments is this:
[
  {
    "m_cloneFrom": "Mistlands_thunder",
    "m_name": "Clear Summer Ocean",
    "m_isWet": false,
    "m_isFreezing": false,
    "m_isFreezingAtNight": false,
    "m_isCold": true,
    "m_isColdAtNight": true,
    "m_alwaysDark": false,
    "m_ambientList": "",
    "m_musicMorning": "",
    "m_musicEvening": "",
    "m_musicDay": "LightRainDay",
    "m_musicNight": ""
  }
]
Now I created a non vanilla folder called ClearOceanDay in the Custom Music folder
[
  {
    "m_cloneFrom": "Mistlands_thunder",
    "m_name": "Clear Summer Ocean",
    "m_isWet": false,
    "m_isFreezing": false,
    "m_isFreezingAtNight": false,
    "m_isCold": true,
    "m_isColdAtNight": true,
    "m_alwaysDark": false,
    "m_ambientList": "",
    "m_musicMorning": "",
    "m_musicEvening": "",
    "m_musicDay": "ClearOceanDay",
    "m_musicNight": ""
  }
]
I dropped the same mp3 in there, run the game again, changed the env.... and the weather changes but the music is not playing. 
I checked Aedenthorn's code for custom audio and it seems that it should be loading everything including the custom stuff,  and then it proceeds into assigning the loaded stuff as appropriately to the registered environments.
(the code is here: https://github.com/aedenthorn/ValheimMods/blob/master/CustomAudio/BepInExPlugin.cs)

So this new Clear Summer Ocean should in theory play the music that is in the custom folder just fine. 
I hope this is not another "comma case" somewhere but I do not think so, changing the text from ClearOceanDay to LightRainDay works. It must be something else.

How did you add music to the frozen ocean? 
Basically what I am trying to achieve at the moment is to change the music more appropriately based on the season. Meadows in summer having different music or sounds than Meadows in Winter.

**Comment by shudnal** [comment]
I have no idea would it work with custom audio only. https://thunderstore.io/c/valheim/p/JereKuusela/Expand_World_Music/ should work better.

**Comment by shudnal** [comment]
I've added music to frozen ocean by duplicating Intro music and setting low volume for better effect

**Comment by gotonickpappas** [comment]
The Expand World Music would work with the environments added from Seasons?!
I will look into it. Thank you so much for your help! :)

**Comment by gotonickpappas** [comment]
The Expand World Music requires the Expand World Data, which creates a yaml file "expand_environments.yaml"
In that file there are parameters defined for environments which I think would override the ones from Seasons. 
Example for Clear:
- name: Clear
  isDefault: true
  isColdAtNight: true
  windMin: 0.1
  windMax: 0.6
  ambColorDay:
    r: 0.463
    g: 0.574
    b: 0.706
    a: 1
........
  fogDensityDay: 0.003
  lightIntensityDay: 1.7
  lightIntensityNight: 1
  sunAngle: 45
  
  So I am not sure how to go about integrating Seasons with the expand world data/music. I would like to have the settings exactly as you have them for seasons as they are quite incredible.
  Were you able to add custom music to a particular custom environment in the past?

**Comment by shudnal** [comment]
Ok I'll check on how Custom Audio manages music

**Comment by shudnal** [comment]
It looks like Custom Audio do some extra unneeded step aside from music registering. 

I don't want to add custom music to Seasons as I have no clip to use as an example.

I think you should stick to Expand World Music + Expand World Data with two options:
* add music with EWM and add environment with Seasons, setting music name as your custom music
* add music with EWM and environment with EWD and then use that environment name in Custom Biome Environments.json

I think first option is more safe and stable but second worth trying if it's something with first one.

**Comment by gotonickpappas** [comment]
It does not look that it is possible to use EWM as it does not detect or simply ignores the environments added from Seasons. 
So as an example, with Seasons you add Clear Winter and Clear Summer as two distinct new environments. 
However Expand World lists only these in the expand_environments.yaml:
<img width="295" alt="image" src="https://github.com/shudnal/Seasons/assets/18703679/d1df24c3-0e5e-493a-a0d3-cc97f56cd537">

Similarly it has different settings that those from Seasons in the expand_biomes.yaml:
<img width="302" alt="image" src="https://github.com/shudnal/Seasons/assets/18703679/d33cc6a1-db5d-43c6-88d1-b3a5639ec705">

These appear to be the default vanilla environments with their default parameters.
EWM cannot be used without the Expand World Data which is what is introducing these yaml files, with the default settings. 

So, it is not clear to me if the two (Expand World and Seasons) can work together out of the box. It appears that it is not possible. 
Your second suggestion (adding environment with EWD and then add that to the Custom Biome Environments.json) is probably going to work but that means that the user has to simply recreate the environments you have already made in EWD. 

Seasons btw is simply fantastic, and with what I have achieved with Custom Music so far, the results are unbelievably cinematic. Simply incredible stuff.

**Comment by shudnal** [comment]
Isn't there an option to only add custom music and don't use any other EWD features? Seasons and EWD should be compatible in general as I had to do multiple compatibility improvements.

**Comment by shudnal** [comment]
What is the license of the music you use? If the tracks is available for non commercial public usage I could use it in some way like custom evening music as an example. Implementing custom music only for that limited purpose is kinda easy btw.

**Comment by gotonickpappas** [comment]
No, if there is a way to not use the rest of EWD I am unaware of it. As you can see on the screenshots, the expand_environments.yaml from EW does not seem to detect the changes introduced by Seasons. 
It should be listing the new environments but it does not. This could be a loading order issue although that would be the first time I would have seen something that has to do with loading order in Valheim. So if it does not detect the new environments and the yaml files override your changes, (or seasons overrides the EWD ones. I am not sure). 

There are countless copyright free tracks that one can use online for example:
https://www.youtube.com/watch?v=j_sTMAIK4OU&list=PLfP6i5T0-DkKJPd8anaZI_HIEiAbIEO8u&index=2

I am not using those though. I am using copyrighted music that I own, since it is for personal use. I am not sure that licensing is an issue as long as a mod provides a mechanism for changing the music, but not the content itself. Similar to how you have provided the mechanism for changing textures if you want in Seasons, but you do not provide the textures themselves, so if someone has copyrighted textures that they use for themselves they can do so. 

Yes replacing e.g. the evening or dawn music is easy. 
Replacing music like e.g. the meadows is also easy. 

The problem becomes not easy when you try to do more advanced stuff like, as an example, have different music playing when in Dark Forest and it storms vs when you are in Meadows and it storms. 
Or if you want different tracks for Meadows when its summer vs when its fall or winter which truly creates some incredibly cinematic experience. 
For those cases, the solution is to make new environments and attach the different tracks or collection of tracks to that specific environment.
 
Sailing is another example, where it is easy to simply change the sailing music, but it is tricky to change the sailing during summertime, vs sailing during winter, sailing when it is clear and calm vs sailing when it is stormy. 

If I am not mistaken sailing is an environment on its own, triggered by boat speed. 

As I said, the combination of having seasons with the ability to alter the mood through music creates some jaw dropping moments. 

It seems that the issue with the mechanism provided by Custom Audio is that it appears to not be possible to add a new environment (or biome) folder that can be used by the game. Even though it looks like Custom Audio reads all the folders and adds them to EnvMan. 

With EWM you can add new folders and files just fine and use them IF you make environments using EWD, but if you want to use the environments from Seasons... it wont detect the new environments, which makes for a situation completely equivalent to that of custom audio.

**Comment by shudnal** [comment]
Seasons makes changes after EWD as I alter environments on every season change and EWD only initially.
So EWD is unaware of Seasons' custom environments and changes and henceforth it looks like vanilla.
The logic is to let EWD change environments and other stuff as it likes and then override (if applicable) with seasonal changes.

CustomAudio works mostly the same. On environment manager initialization it replaces some tracks and adds music to list. 
Now you should be able to use that music identifier in Seasons configuration file.

Like you add new music i.e. **niceVikingMusic.mp3**, named it **nice_viking_music** using CustomAudio. 
And then in Custom environments.json you just use m_musicDay: "nice_viking_music" in your new environment and it should work in theory.

**Comment by shudnal** [comment]
Still you made me thinking of adding support of simple custom music extension for use in environments. I already parse used audio clips to be able to swap environments. Adding some custom music and ambient tracks looks pretty easy. And if some client doesn't have that track in their folder nothing will break. Music just won't be playing.

**Comment by gotonickpappas** [comment]
> The logic is to let EWD change environments and other stuff as it likes and then override (if applicable) with seasonal changes.

This makes a ton of sense actually. Well done. 

Now I agree that in theory Custom Audio should be able to do exactly what you described but it does not. 
<img width="1099" alt="image" src="https://github.com/shudnal/Seasons/assets/18703679/bc9955a8-a953-443a-86a3-917464cb2bef">
So if you put new music in a folder that corresponds to a vanilla environment (green arrow) then all is fine. It plays the music. 
If on the other hand you want to create a new folder to put your new music then it does not. (red arrow). 
Remember the goal here is to separate the environments so that LightRainDay becomes LightRainDaySummer, LightRainDayFall, etc, so the new folders are needed.
I do not understand why this is so as it is clear from its code that Custom Audio reads all the files in the folders under Music and adds them the the env manager. It is probably a Unity thing, that I am not familiar with. 

Yes you are right. If the music is not in the folder, then the default plays, or if it is custom (as in the EWM) and is missing then nothing plays. 

In my opinion you should indeed consider adding this functionality so that the changes in music can be achieved "natively" in seasons, instead of having to rely on other mods. It appears to be a non trivial but also not too difficult task in the sense that you do not really have to reinvent the wheel on this matter, and I know that I probably sound like a broken record with this... but the combination of the visuals with music is quite indescribable. 
You have done some spectacular job.
Also it runs amazingly well on all my kids' potato pcs. Basically, as if it is not there.

**Comment by shudnal** [comment]
Thing is EWM works fine but with EWD as dependancy it becomes a nuisance.
Custom Audio is fine but overall mod purpose is music replacing not just adding.
I need some time to think how it should be done. But I decided to make it possible.

Performance wise I usualy play with RenderLimits and distance objects at 12 which is immersive but it's always like 12000 trees around. In that setup my performance is hit by 5-8% like 160->150fps. I made everything I could to make it performant. It took about 3 months in general but the result speaks for itself.

**Comment by shudnal** [comment]
One of the features I think of is also custom sailing music for different environments. Right now we have only **sailing** and **sailing_ashlands** which I think could be extended to use more conditions like day/night/storm and so.

Right now I'm thinking of combining features of EWM and CustomAudio into distinct mod with additional features.
And in Seasons will be only some baseline support for custom music tracks for environments. With additional support of seasonal variants of different musics later.
I hardly believe it will be used by more than one enthusiast but it will still be worth it.

**Comment by gotonickpappas** [comment]
Custom Audio can both replace and add music and/or sounds. It is pretty great. I think ewm can do the same thing but unlike Custom Audio, you have to manually enter the file names and paths in the yaml after you create the folders and copy the files. 

Great news about making it possible. It is going to be awesome! 

I too play with RenderLimits at 12 and I do not have an impact. My kids though have potato laptops all four of them, and they play with the default distances. However, as I said with the default values and Seasons, there is no impact at all. 

Sailing music is something that I do not understand how it is that they did not split it in parts in the first place. Epic stuff can be made from it. Sailing in daylight in sunlight with a calm sea should be very different than struggling in a thunderstorm at night in the middle of the ocean. If we were scoring a movie we would be scoring the situations very differently.

I am not sure how exactly you could split the sailing into more variants as it is true that you can create a new sail environment but they all trigger from boat speed which would be common and it is my impression that this is hard coded. Perhaps I am wrong. 

I disagree about the popularity of the custom music per season/weather/daynight feature. Once it's there it will be used because some jaw dropping cinematic moments happen with this and people will take notice on YouTube. 
Basically, the game becomes significantly more like a movie and the Seasons contributes a lot to that feeling. Not just graphically but gameplay wise as well. Winter is not just pretty to look at, it makes an impact in how you play and the music can intensify that feeling a great deal.

**Comment by shudnal** [comment]
That's actually should be pretty elegant and easy.
![image](https://github.com/shudnal/Seasons/assets/20457922/c0782c3d-5e27-4cfd-8f1b-e6eb88243c59)
In postfix to that method we just add check for our condition and just override the result if we have custom music.
But as usual implementing is easy but configuring that shit is another pain in the ass.

By unpopularity I meant it's harder to make good. It's not just pick epic music track and paste it here and voila. It should be looped at least. The dynamic range should be limited to not bother player by loud music. It's doable of cource but it takes some enthusiasm.

**Comment by shudnal** [comment]
Music override will be done by distinct mod. And on top of that and other places where music is used we can just add seasonal variants by adding postfix to music name.
Like in that sailing example. We read the result of the method. And it is just usual **sailing**. We look into our configured seasonal music variant, we found music override "sailing" and just replace result with **sailing_summer**. Which is already added in music list. Voila we have seasonal variants for any music.
And again the ease of configurating is major pain in the ass.

**Comment by gotonickpappas** [comment]
Ah yes, of course, scoring, mixing, looping and so forth is an art in itself. No argument there. I have half done it and I can attest to that. It is definitely not a pick a few epic music tracks and there you go. In fact often you need to add stuff that the player does not notice them being there. But once you do find the right stuff, truly epic moments emerge.

The thing is that right now there is no  framework in place for doing it, so one has to struggle both with the doing all that stuff but also with the simple loading of the material for testing what would sound good, how it should be mixed, looped etc. 

So the way it is implemented right now is that they check in the biome is AshLands in the GetSailingMusic method, and if it is not they just return sailing? Gee... so that if could be replaced with a switch case for other biomes?

**Comment by shudnal** [comment]
Yep that's right. We can check for any condition we need and come up with best fitting music for that given moment.

**Comment by shudnal** [comment]
I think it's better to continue in Discord. https://discord.com/users/shudnal
That issue has become irrelevant anyway. You made your custom environment working. And adding custom music support is another story.

**Comment by gotonickpappas** [comment]
Issue resolved, as there was no issue really.
